_posts/2019-08-23-unet/unet.Rmd

```diff
@@ -0,0 +1,437 @@
+
+In practice, the classic architectures use (max) pooling or convolutions with `stride` > 1 to achieve those successive abstractions -- necessarily resulting in decreased spatial resolution.
+So how can we use a convnet and still preserve detail information? In their 2015 paper _U-Net: Convolutional Networks for Biomedical Image Segmentation_ [@RonnebergerFB15], Olaf Ronneberger et al. came up with what four years later, in 2019, is still the most popular approach. (Which is to say something, four years being a long time, in deep learning.)
+
+The idea is stunningly simple. While successive encoding (convolution / max pooling) steps, as usual, reduce resolution, the subsequent decoding -- we have to arrive at an output of  size same as the input, as we want to label every pixel! -- does not simply upsample from the most compressed layer. Instead, during upsampling, at every step we feed in information from the corresponding, in resolution, layer in the downsizing chain.
+
+For U-Net, really a picture says more than many words:
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "U-Net architecture from Ronneberger et al. 2015."}
+knitr::include_graphics("images/unet.png")
+```
+
+
+Specifically, at each upsampling stage we _concatenate_ the output from the previous layer with that from its pendant in the compression stage. The final output is a _mask_ of size the original image, obtained via 1x1-convolution; no final dense layer is required, instead the output layer is just a convolutional layer with a single filter. 
+
+Now let's actually train a U-Net. We're going to use the [`unet` package ](https://github.com/r-tensorflow/unet) that lets you create a well-performing model in a single line:
+
+```{r}
+remotes::install_github("r-tensorflow/unet")
+library(unet)
+
+# takes additional parameters, including number of downsizing blocks, 
+# number of filters to start with, and number of classes to identify
+# see ?unet for more info
+model <- unet(input_shape = c(128, 128, 3))
+```
+
+So we have a model, and it looks like we'll be wanting to feed it 128x128 RGB images. Now how do we get these images?
+
+## The data
+
+To illustrate how applications arise even outside the area of medical research, we'll use as an example the Kaggle [Carvana Image Masking Challenge](https://www.kaggle.com/c/carvana-image-masking-challenge/overview). The task is to create a segmentation mask separating cars from background. For our current purpose, we only need `train.zip` and `train_mask.zip` from the [archive provided for download](https://www.kaggle.com/c/6927/download-all). In the following, we assume those have been extracted to a subdirectory called `data_raw`.
```

Comment:
Have you considered making a repo with the data and code for the project? Was just thinking it could be nice to link to something that already has the directory structure, etc. you're describing.

Reply:
the data is 24 G :-(
but for the code, you're right, we have the same code in a vignette, I can add a footnote to say so, then people don't have to copy-paste